lib/std/crypto/sha3.zig

```diff
@@ -196,6 +307,242 @@ fn ShakeLike(comptime security_level: u11, comptime delim: u8, comptime rounds:
+            context: ?[]const u8 = null,
+        };
+
+        /// Initialize a state for the TupleHash function, with an optional context.
+        /// If the context is going to be reused, the structure can be initialized once, and cloned for each message.
+        /// This is more efficient than reinitializing the state for each message at the cost of a small amount of memory.
+        ///
+        /// A key can be optionally added to the context to create a keyed TupleHash function, similar to KMAC.
+        pub fn initWithOptions(options: Options) Self {
+            const cshaker = CShaker.init(.{ .context = options.context });
+            return Self{
+                .cshaker = cshaker,
+            };
+        }
+
+        /// Initialize a state for the MAC function.
+        pub fn init() Self {
+            return initWithOptions(.{});
+        }
+
+        /// Add data to the state, separated from previous updates.
+        pub fn update(self: *Self, b: []const u8) void {
+            const encoded_b_len = NistEncodedLength.encode(b.len, .left);
+            self.cshaker.update(encoded_b_len.slice());
+            self.cshaker.update(b);
+        }
+
+        /// Return an authentication tag for the current state.
+        pub fn final(self: *Self, out: []u8) void {
+            const encoded_out_len = NistEncodedLength.encode(out.len, .right);
+            self.cshaker.update(encoded_out_len.slice());
+            self.cshaker.final(out);
+        }
+
+        /// Align the input to a block boundary.
+        pub fn fillBlock(self: *Self) void {
+            self.cshaker.fillBlock();
+        }
+
+        /// Squeeze a slice of bytes from the state.
+        /// `out` can be any length, and the function can be called multiple times.
+        pub fn squeeze(self: *Self, out: []u8) void {
+            if (!self.xof_mode) {
+                const encoded_out_len = comptime NistEncodedLength.encode(0, .right);
+                self.update(encoded_out_len.slice());
+                self.xof_mode = true;
+            }
+            self.cshaker.squeeze(out);
+        }
+
+        pub const Error = error{};
+        pub const Writer = std.io.Writer(*Self, Error, write);
+
+        fn write(self: *Self, bytes: []const u8) Error!usize {
+            self.update(bytes);
+            return bytes.len;
+        }
+
+        pub fn writer(self: *Self) Writer {
+            return .{ .context = self };
+        }
+    };
+}
+
+/// The NIST encoded length format.
+pub const NistLengthEncoding = enum { left, right };
+
+/// A length encoded according to NIST SP 800-185.
+pub const NistEncodedLength = struct {
+    /// The size of the encoded value, in bytes.
+    len: usize = 0,
+    /// A buffer to store the encoded length.
+    buf: [@sizeOf(usize) + 1]u8 = undefined,
+
+    /// Encode a length according to NIST SP 800-185.
+    pub fn encode(len: usize, comptime encoding: NistLengthEncoding) NistEncodedLength {
```

Comment:
This function should move into `NistLengthEncoding` namespace and have the argument order flipped.

Reply:
Done.